server/services/analyticsServiceUtils.ts

```diff
@@ -36,6 +36,61 @@ export function mapRowsAndSumTotals<
   return rows
 }
 
+/**
+ * Ensures that months without data are still present for last 12 months (excluding this month)
+ */
+export function addMissingMonths(rows: TrendsReportRow[], valueCount: number) {
+  const month = new Date()
+  month.setDate(1)
+  month.setHours(12, 0, 0, 0)
+  for (let monthsAgo = 1; monthsAgo <= 12; monthsAgo += 1) {
+    month.setMonth(month.getMonth() - 1)
+    if (
+      !rows.some(
+        ({ month: someMonth }) =>
+          someMonth.getFullYear() === month.getFullYear() && someMonth.getMonth() === month.getMonth()
+      )
+    ) {
+      rows.push({
+        month: new Date(month),
+        values: Array(valueCount).fill(0),
+        total: 0,
+        population: 0,
+      })
+    }
+  }
+  rows.sort(compareMonths)
+}
+
+/**
+ * Removes data that is not from last 12 months (excluding this month)
+ */
+export function removeMonthsOutsideBounds(rows: TrendsReportRow[]): TrendsReportRow[] {
+  const previousMonth = new Date()
+  previousMonth.setDate(1)
+  previousMonth.setHours(12, 0, 0, 0)
+  previousMonth.setMonth(previousMonth.getMonth() - 1)
+  const latestYear = previousMonth.getFullYear()
+  const latestMonth = previousMonth.getMonth()
+
+  const yearAgo = new Date()
+  yearAgo.setDate(1)
+  yearAgo.setHours(12, 0, 0, 0)
+  yearAgo.setMonth(yearAgo.getMonth() - 12)
+  const oldestYear = yearAgo.getFullYear()
+  const oldestMonth = yearAgo.getMonth()
+
+  return rows.filter(({ month: rowMonth }) => {
+    const year = rowMonth.getFullYear()
+    const month = rowMonth.getMonth()
+    return (
+      (year > oldestYear && year < latestYear) ||
+      (year === oldestYear && month >= oldestMonth) ||
+      (year === latestYear && month <= latestMonth)
+    )
```

Comment:
I'm struggling to follow this a bit tbh. I guess we rely on the fact we don't have future months (e.g. or `(year > oldestYear && year < latestYear)` would mean something like July 2023 would be in rows).

Edge case aside, I'm wondering if it would be easier to use `Date` comparisons? e.g. _something like_:

```javascript
  const latestMonth = new Date()
  latestMonth.setDate(1)
  latestMonth.setHours(12, 0, 0, 0)
  latestMonth.setMonth(latestMonth.getMonth() - 1)

  const earliestMonth = new Date()
  earliestMonth.setDate(1)
  earliestMonth.setHours(12, 0, 0, 0)
  earliestMonth.setMonth(earliestMonth.getMonth() - 12)

  return rows.filter(({ month: rowMonth }) => {
    return rowMonth >= earliestMonth && rowMonth <= latestMonth
  })
```

or something like that.

Reply:
i don't think date comparisons would be a good idea because of time zones. perhaps i could map the YEAR-MONTH strings into `Date`s _after_ fixing the timeline.